### PR TITLE
misc fixes for heterogeneous cluster support

### PR DESCRIPTION
**Pull request by ggouaillardet**
Refs. open-mpi/ompi#2838

### CODE REVIEW

File: opal/datatype/opal_convertor.c [hunk 1]
```diff
@@ -470,7 +470,8 @@ int32_t opal_convertor_set_position_nocheck( opal_convertor_t* convertor,
             }                                                             \
         }                                                                 \
         convertor->remote_size *= convertor->count;                       \
-        convertor->use_desc = &(datatype->desc);                          \
```
Comment (bosilca) [inline opal/datatype/opal_convertor.c]: At this point the convertor use_desc points to the optimized description, which is a description without datatype information (aka. not suitable for heterogeneous operations). I do think that forcing the switch to the default description (which contains the datatype information) is the right thing here. However, I would remove this line from the OPAL_CONVERTOR_COMPUTE_REMOTE_SIZE macro and instead put it right in the ```if( ((convertor->flags &...``` in the OPAL_CONVERTOR_PREPARE macro.

**Comment by hppritcha** [comment]
SS Botany Bay OS-X system doesn't like this PR:

```
08:36:37   CC       opal_copy_functions_heterogeneous.lo
08:36:37 opal_copy_functions_heterogeneous.c:18:10: fatal error: 'ieee754.h' file not found
08:36:37 #include <ieee754.h>
08:36:37          ^
08:36:37 1 error generated.
08:36:37 make[2]: *** [opal_copy_functions_heterogeneous.lo] Error 1
08:36:37 make[1]: *** [install-recursive] Error 1
```

**Comment by ggouaillardet** [comment]
:bot:mellanox:retest

**Comment by ibm-ompi** [comment]
The IBM CI (PGI Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/8281105948edfccf8ea59af74273bbb1

**Comment by ibm-ompi** [comment]
The IBM CI (GNU Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/28eea78fe6bd6968a2b5c9bc31d2b848

**Comment by ibm-ompi** [comment]
The IBM CI (GNU Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/35006c2fefb19f75a7f9b01101d61267

**Comment by ibm-ompi** [comment]
The IBM CI (XL Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/0cabf9fbf21623f5d78e0576a6a69de7

**Comment by ggouaillardet** [comment]
@bosilca i updated the PR (there were quite some changes in datatype handling)

could you please give it a final review before i merge ?

File: opal/datatype/opal_convertor.c [hunk 2]
```diff
@@ -474,14 +474,16 @@ opal_datatype_compute_remote_size( const opal_datatype_t* pData,
...
-size_t opal_convertor_compute_remote_size( opal_convertor_t* pConvertor )
+size_t opal_convertor_compute_remote_size( opal_convertor_t* pConvertor, bool send )
 {
     opal_datatype_t* datatype = (opal_datatype_t*)pConvertor->pDesc;
     
     pConvertor->remote_size = pConvertor->local_size;
     if( OPAL_UNLIKELY(datatype->bdt_used & pConvertor->master->hetero_mask) ) {
         pConvertor->flags &= (~CONVERTOR_HOMOGENEOUS);
-        pConvertor->use_desc = &(datatype->desc);
+        if (!(send && pConvertor->flags & OPAL_DATATYPE_FLAG_CONTIGUOUS)) {
```
Comment (bosilca) [inline opal/datatype/opal_convertor.c]: This information should already be in the convertor flags (CONVERTOR_SEND).

**Comment by ggouaillardet** [comment]
@bosilca i made the requested changes.
the ```ddt_pack_hetero``` test has been removed since this PR was created, and ```unpack_hetero``` works regardless ```--enable-heterogeneous``` is set or not

**Comment by bosilca** [comment]
One last question. I noticed you changed the prototypes of the PMPI functions in ompi/mpi/fortran/mpif-h/prototypes_mpi.h. Does this change breaks our ABI ?

File: test/datatype/unpack_hetero.c [hunk 3]
```diff
@@ -1,6 +1,6 @@
 /* -*- Mode: C; c-basic-offset:4 ; -*- */
 /*
- * Copyright (c) 2014-2016 Research Organization for Information Science
+ * Copyright (c) 2014-2017 Research Organization for Information Science
```
Comment (jsquyres) [inline test/datatype/unpack_hetero.c]: Did you change anything in this file?

Comment (ggouaillardet) [inline test/datatype/unpack_hetero.c]: i will double check that, maybe there used to be a fix that has already been merged or made obsolete by a revamp.

**Comment by ggouaillardet** [comment]
@bosilca, the changes involves the prototype of a function pointer.
so there are several ways to see this.
1. this breaks ABI
2. this does not break ABI (the pointer size did not change)
3. current implementation is broken anyway, so this really a bug fix, and since the pointer size did not change, this is an acceptable commit

i'd rather go with the latter option.

**Comment by bosilca** [comment]
We've been extremely careful not to break the ABI in the middle of a series. I would also tend to go with your latter option, but the RM should be aware of the possible ABI divergence.

**Comment by ggouaillardet** [comment]
got it, i will fix the unnecessary copyright change pointed by @jsquyres and merge this tomorrow into ```master```. then i will PR to ```v3.0.x``` so we will hopefully have this ready for ```v3.0.0```
then i will think if these changes are needed for the ```v2``` branches

**Pull request merged**